### ISSUE TITLE
$config['skins_allowed'] doesn't trump existing skins user preference

### ISSUE THREAD

**Issue by dpc22** [roundcube/roundcubemail#7080]
(Not sure whether this should be a pull request rather than an issue).

The pull request which added the 'skins_allowed' configuration setting https://github.com/roundcube/roundcubemail/pull/6483 includes:

> Currently the part is missing that enforces the skin selection when the "current" skin is not on the  list of available skins and fall back to the default one.

I attach a patch which implements that part, as this is something that I personally need. It is less clear to me whether this is something that other people want.

This works nicely for me. The only bit that I'm not sure about is whether:

    $myskin = $this->app->default_skin;

is an approved means of getting the initial value of $config['skin'] before user preferences are loaded.

[rcube_output_html.patch.txt](https://github.com/roundcube/roundcubemail/files/3896603/rcube_output_html.patch.txt)

**Comment by alienmindbender** [comment]
I think this should not be the default behaviour.
We restricted the selection to elastic and larry.  Some users still use classic and set this as preference before the introduction of elastic. This would enforce elastic on them (which is our default). New users however should not have the choice to choose classic.

If you wan't to enforce a specific skin, why not use moduserprefs.sh?

**Comment by dpc22** [comment]
Maybe two separate lists for "enabled_skins" and "allowed_skins"? "allowed_skins" is a slightly unfortunate choice of name when it currently actually means "user_selectable_skins".

We are currently in the process of migrating from a custom skin to (very lightly tweaked) versions of larry and elastic which have a replacement login page. I appreciate that this is rather unusual: I just need a fall back plan in case we run into problems. moduserprefs.sh is a little scary with 40k users.

**Comment by alecpl** [comment]
We don't need another config option. I agree this should have been enforced from the beginning. Imo, this is most-expected case.

**Comment by alienmindbender** [comment]
I agree there is no need for another config option.
But I'am not super happy with the enforcement on the user and think there is no need to go that route, since all the tools to do that are already there. Furthermore you'll obsolete the users skin preferences within the database in this particular example. The clean way - imho - would be to set a default skin, fordbid user selection and delete "skin" from the users preferences.
I (and probably other too) will have to force migrate users then - I don't really mind, since it's only a handfull of accounts. But as I said - the possibility to enforce a skin are already there and changing the behaviour of allowed_skins would just strip some flexibility away.

**Comment by alecpl** [comment]
Just think why would you want to limit skins with `skins_allowed` option? The only sensible reason I can think of is to disable access to some skins (e.g. classic, not supported by all plugins, or disable all except your custom skins) without removing skin folders. So "allowing users to use a different skin that the one in `skins_allowed` sounds even like a bug. But maybe you know other use-cases?

**Comment by alecpl** [comment]
Additional note, why not removing skin folders is so important? Because your allowed skins might depend on them (skin inheritance functionality).

**Comment by alienmindbender** [comment]
Just to slowly deprecate a skin. New users should not be allowed to select deprecated ones - existing users can still use it for the period defined by the administrator. 
I used the option, since it clearly states its behaviour in default.php:
`limit skins available/shown in the settings section`
Because of that it doesn't seem like a bug to me, but the intended behaviour - but I have to agree that the naming of the directive is a little bit misleading.

**Comment by alecpl** [comment]
@thomascube what do you think?

**Comment by thomascube** [comment]
Despite the explanation in the config comment, I'd expect `skins_allowed` to also affect the actual selection (i.e. saved skin). I honestly don't see that use-case where one wants to remove a skin from the selection but still kinda allow it. It means that whenever a user visits the preferences page and save it she'll loose their deprecated skin selection without further notice.

And we definitely don't want yet another option for freak cases. Therefore I suggest to change the meaning and behavior of `skins_allowed`.

**Comment by alienmindbender** [comment]
I wouldn't exactly call this a "freak case"; I just don't want to force migrate users (right now). And I don't want more users to select a skin which won't be supported in the future (by you/by us)... I probably would have to deal with more users then...
Changing that option would force me to deprecate the skin for good right now - or I have to stall the next update for the time beeing.
Despite the naming of the directive, the explanation was pretty clear on what I had to expect when using `skins_allowed`; I don't think there was a lot of wiggle room - interpretation wise...

But:
> It means that whenever a user visits the preferences page and save it she'll loose their deprecated skin selection without further notice.

I did thought about that and while technically the user would be responsible for the change, that  would be actually as bad as enforcing a skin. I dismissed that thought to early - you're right.
